### PR TITLE
HWKAPM-487 Define generic cache API with checked exception on store. …

### PR DESCRIPTION
**Pull request by objectiser**
…Removed unused support for retrieving multiple consumer information, which will be handled later in jira hwkapm 356.

### CODE REVIEW

**Comment by objectiser** [comment]
Review notes:
- related to adding RetryAttemptException as explicit trigger for retrying failed event processing
- storing events in cache previously didn't have a checked exception, so if the cache was to fail, then would not trigger the retry.
- took opportunity to define a generic cache API to support store with checked exception - previously had specific cache APIs as (for example) the communication details were retrieved differently depending on whether there are multiple consumers - however this mechanism hasn't been fully implemented yet, so instead simplifies API to be generic for main use, and will then extend the CommunicationDetailsCache when handling multiple consumers.

**Comment by objectiser** [comment]
@metlos Sorry one more for you, as its related to the RetryAttemptException and the specific case you picked up on related to the cache not having a checked exception on the store method.

File: server/api/src/main/java/org/hawkular/apm/server/api/services/Cache.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * @param <T> The type of cached information
+ */
+public interface Cache<T> {
+
+    /**
+     * This method retrieves the information based on the supplied
+     * id.
+     *
+     * @param tenentId The tenant id
```
Comment (metlos) [inline server/api/src/main/java/org/hawkular/apm/server/api/services/Cache.java]: s/tenentId/tenantId

File: server/api/src/main/java/org/hawkular/apm/server/api/services/Cache.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+     * @return The information, or null if not found
+     */
+    T get(String tenantId, String id);
+
+    /**
+     * This methods stores the information in the cache. The
+     * id for each piece of information is obtained from the information
+     * itself.
+     *
+     * @param tenentId The tenant id
```
Comment (metlos) [inline server/api/src/main/java/org/hawkular/apm/server/api/services/Cache.java]: s/tenentId/tenantId

File: server/api/src/main/java/org/hawkular/apm/server/api/task/ProcessingUnit.java [hunk 3]
```diff
@@ -106,7 +106,12 @@ public void handle(String tenantId, List<T> items) throws Exception {
         List<T> retries = null;
         RetryAttemptException lastException = null;
 
-        processor.initialise(tenantId, items);
+        try {
+            processor.initialise(tenantId, items);
+        } catch (RetryAttemptException e) {
+            retries = items;
```
Comment (metlos) [inline server/api/src/main/java/org/hawkular/apm/server/api/task/ProcessingUnit.java]: How does the retry handler distinguish "failure in initialization + failure in handling all items" from "success in initialization + failure in handling all items"?

Comment (objectiser) [inline server/api/src/main/java/org/hawkular/apm/server/api/task/ProcessingUnit.java]: It doesn't really need to - whether the initialisation fails, or the processing of all the events (as a batch) fails, the end result is that the events need to be retried.

Comment (objectiser) [inline server/api/src/main/java/org/hawkular/apm/server/api/task/ProcessingUnit.java]: Sorry just realised what you mean  - it currently would process the events if init failed, which it shouldn't. I'll fix that.

**Comment by metlos** [comment]
:+1: ship it :smiley:

**Comment by objectiser** [comment]
@metlos Thanks!

**Pull request merged**